docs/gallery/howto/autogen/graph_builder.py

```diff
@@ -55,28 +36,78 @@ def multiply(x, y):
     return x * y
 
 
-# use task.graph_builder decorator, expose the output of "multiply" task
-# as the output of the `WorkGraph`.
+def add_multiply(x, y, z):
+    wg = WorkGraph()
+    wg.add_task(add, name="add", x=x, y=y)
+    wg.add_task(multiply, name="multiply", x=z)
+    wg.add_link(wg.tasks["add"].outputs[0], wg.tasks["multiply"].inputs["y"])
+    return wg
+
+
+wg = WorkGraph("nested_workgraph")
+add_multiply1 = wg.add_task(add_multiply(x=Int(2), y=Int(3), z=Int(4)))
+wg.to_html()
+
+# %%
+# Run the workgraph
+
+wg.run()
+
+# %%
+# However linking the two WorkGraphs will not work
+
+wg = WorkGraph("nested_workgraph")
+add_multiply1 = wg.add_task(add_multiply(x=Int(2), y=Int(3), z=Int(4)))
+
+try:
+    wg.add_task(
+        add_multiply(x=add_multiply1.outputs["multiply.result"], y=Int(3), z=Int(4))
+    )
+except Exception as err:
+    print(err)
```

Comment:
Actually, one can link the two WorkGraphs. But we need to write the code in a different way:
```python
def add_multiply(x=None, y=None, z=None):
    wg = WorkGraph()
    wg.add_task(add, name="add", x=x, y=y)
    wg.add_task(multiply, name="multiply", x=z)
    wg.add_link(wg.tasks["add"].outputs[0], wg.tasks["multiply"].inputs["y"])
    return wg

wg = WorkGraph("nested_workgraph")
add_multiply1 = wg.add_task(add_multiply(x=2, y=3, z=4), name="add_multiply1")
add_multiply2 = wg.add_task(add_multiply(y=5, z=6), name="add_multiply2")
wg.add_link(add_multiply1.outputs["multiply.result"], add_multiply2.inputs["add.x"])
wg.run()
```

The difference between the above code and a `graph_builder` task is that in `graph_builder`, the workgraph is dynamic and only created during execution, while the above workgraph is static, so that we can access the socket directly, e.g., `add.x`.
Both can used to create nested workgrpah, but `graph_builder` is a black box, which is the downside. The upside is that it allows dynamic workgraph generation.

Reply:
Okay used that example and explained a bit

Reply:
Hm.. it is not properly working
```python
# define add task
@task.calcfunction()
def add(x, y):
    return x + y


# define multiply task
@task.calcfunction()
def multiply(x, y):
    return x * y


def add_multiply(x=None, y=None, z=None):
    wg = WorkGraph()
    wg.add_task(add, name="add", x=x, y=y)
    wg.add_task(multiply, name="multiply", x=z)
    wg.add_link(wg.tasks["add"].outputs[0], wg.tasks["multiply"].inputs["y"])
    return wg


wg = WorkGraph("nested_workgraph")
# Creating a task from the WorkGraph
add_multiply1 = wg.add_task(add_multiply(x=Int(2), y=Int(3), z=Int(4)))
add_multiply2 = wg.add_task(add_multiply(x=Int(2), y=Int(3)))
# link the output of a task to the input of another task
wg.add_link(add_multiply1.outputs[0], add_multiply2.inputs["multiply.x"])
wg.to_html()

# %%
# Run the workgraph

wg.run()
```
Gives
```
Error: Error in task multiply: Cannot convert value of type <class 'aiida.orm.utils.managers.NodeLinksManager'> to AiiDA type.
```

Reply:
Okay now it works!